editor/src/clj/editor/app_view.clj

```diff
@@ -753,32 +754,174 @@
         (error-reporting/report-exception! error)
         nil))))
 
-(defn async-build! [project evaluation-context prefs {:keys [debug? engine?] :or {debug? false engine? true}} old-artifact-map render-build-progress! result-fn]
-  (assert (not @build-in-progress?))
-  (reset! build-in-progress? true)
-  (future
-    (try
-      (render-build-progress! (progress/make "Building project..." 1))
-      (let [extra-build-targets (when debug?
-                                  (debug-view/build-targets project evaluation-context))
-            build-results (ui/with-progress [_ render-build-progress!]
-                            (build-project! project evaluation-context extra-build-targets old-artifact-map render-build-progress!))
-            [engine build-engine-exception] (when (and engine? (nil? (:error build-results)))
-                                              (try
-                                                (ui/with-progress [_ render-build-progress!]
-                                                  (render-build-progress! (progress/make-indeterminate "Fetching Engine"))
-                                                  [(engine/get-engine project evaluation-context prefs (engine/current-platform)) nil])
-                                                (catch Throwable e
-                                                  [nil e])))]
-        (ui/run-later
-          (try
-            (when result-fn (result-fn build-results engine build-engine-exception))
-            (finally
-              (reset! build-in-progress? false)))))
-      (catch Throwable t
-        (reset! build-in-progress? false)
-        (render-build-progress! progress/done)
-        (error-reporting/report-exception! t)))))
+(defn- cached-build-target-output? [node-id label evaluation-context]
+  (and (= :build-targets label)
+       (project/project-resource-node? node-id evaluation-context)))
+
+(defn- update-system-cache-build-targets! [evaluation-context]
+  ;; To avoid cache churn, we only transfer the most important entries to the system cache.
+  (let [pruned-evaluation-context (g/pruned-evaluation-context evaluation-context cached-build-target-output?)]
+    (g/update-cache-from-evaluation-context! pruned-evaluation-context)))
+
+(defn async-build! [project prefs {:keys [debug? engine?] :or {debug? false engine? true}} old-artifact-map render-build-progress! result-fn]
```

Comment:
The `evaluation-context` argument was removed. We will create new `evaluation-contexts` as needed in order to see changes from build hooks. The entire flow of this function has been rethought and refactored. Hopefully it should be easier to follow the different steps that are performed. Basically each phase does some prep work on the ui thread, moves onto a background thread to do the actual work, then posts a callback onto the ui thread to handle the results. Depending on the results from the background task, it might abort out with an error or launch the next phase which will repeat the process until all the phases have completed successfully.